internal/builder/builder.go

```diff
@@ -1175,11 +1132,102 @@ func (b *Builder) whiteoutLayer(tmpDir string, i int, bpInfo dist.ModuleInfo) (s
 	return fh.Name(), nil
 }
 
-func sortKeys(collection map[string]toAdd) []string {
+func sortKeys(collection map[string]moduleWithDiffID) []string {
 	keys := make([]string, 0, len(collection))
 	for k := range collection {
 		keys = append(keys, k)
 	}
 	sort.Strings(keys)
 	return keys
 }
+
+// splitBuildModules takes the given Build Modules and concurrently read their internal tar files. In
+// case a flattened module is found, it will split each flattened buildpack into an individual module and skip all the empty ones
+// returns an explodedBuildModule array with the Build Module information but also the diffId and the tar file on disk
+func splitBuildModules(kind, tmpDir string, additionalModules []buildpack.BuildModule, logger logging.Logger) ([]moduleWithDiffID, []error) {
+	lids := make([]chan modInfo, len(additionalModules))
+	for i := range lids {
+		lids[i] = make(chan modInfo, 1)
+	}
+
+	for i, module := range additionalModules {
+		go func(i int, module buildpack.BuildModule) {
+			modTmpDir := filepath.Join(tmpDir, fmt.Sprintf("%s-%s", kind, strconv.Itoa(i)))
+			if err := os.MkdirAll(modTmpDir, os.ModePerm); err != nil {
+				lids[i] <- handleError(module, err, fmt.Sprintf("creating %s temp dir %s", kind, modTmpDir))
+			}
+			moduleTars, err := buildpack.ToNLayerTar(modTmpDir, module, logger)
+			if err != nil {
+				lids[i] <- handleError(module, err, fmt.Sprintf("creating %s tar file at path %s", module.Descriptor().Info().FullName(), modTmpDir))
+			}
+			lids[i] <- modInfo{moduleTars: moduleTars}
+		}(i, module)
+	}
+
+	var result []moduleWithDiffID
+	var errs []error
+
+	// maybe we got multiple modules back, we need to skip the empty ones
+	for i, module := range additionalModules {
+		mi := <-lids[i]
+		if mi.err != nil {
+			errs = append(errs, mi.err)
+			continue
+		}
+		for _, moduleTar := range mi.moduleTars {
+			// it could be an individual Buildpack or flattened Buildpack that writes an empty tar on disk
+			eBM := moduleWithDiffID{tarPath: moduleTar.Path()}
+			diffID, err := dist.LayerDiffID(moduleTar.Path())
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "calculating layer diffID for path %s", moduleTar.Path()))
+				continue
+			}
+			if diffID.String() == "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" {
```

Comment:
why is it this exact hash?  is this a magic value that should be turned into a named constant?  do we use it anywhere else?

Reply:
Oh yeah, I forgot to create the constant, the value is the hash of a empty tar file